### PR TITLE
ClusterLoader - Implementing different timers

### PR DESCRIPTION
**Pull request by krzysied**
Tuning sets implementation

### CODE REVIEW

File: clusterloader2/pkg/ticker/timer.go [hunk 1]
```diff
@@ -17,18 +17,54 @@ limitations under the License.
...
+	if ts.QpsLoad != nil || ts.RandomizedLoad != nil {
+		return func() int32 { return int32(1) }, nil
+	}
+	if ts.SteppedLoad != nil {
+		return func() int32 { return ts.SteppedLoad.BurstSize }, nil
+	}
+	return nil, fmt.Errorf("unknown tuningset structure: %v", ts)
+}
+
+func getDelayFunc(ts *api.TuningSet) (func() time.Duration, error) {
```
Comment (shyamjvs) [inline clusterloader2/pkg/ticker/timer.go]: Please add a general comment above explaining what this function is calculating. Or even better if you add a comment above each mode explaining why you compute the delay the way you do it.

File: clusterloader2/pkg/ticker/timer.go [hunk 2]
```diff
@@ -17,18 +17,54 @@ limitations under the License.
...
+	if ts.RandomizedLoad != nil {
+		return func() time.Duration {
+			randomFactor := 2 * rand.Float64()
+			return time.Duration(int(randomFactor * float64(time.Second) / ts.QpsLoad.Qps))
+		}, nil
+	}
+	if ts.SteppedLoad != nil {
+		return func() time.Duration { return ts.SteppedLoad.StepDelay }, nil
+	}
+	return nil, fmt.Errorf("unknown tuningset structure: %v", ts)
```
[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/ticker/timer.go]: done.

File: clusterloader2/pkg/ticker/timer.go [hunk 3]
```diff
@@ -17,18 +17,54 @@ limitations under the License.
...
+	}
+	if ts.SteppedLoad != nil {
+		return func() int32 { return ts.SteppedLoad.BurstSize }, nil
+	}
+	return nil, fmt.Errorf("unknown tuningset structure: %v", ts)
+}
+
+func getDelayFunc(ts *api.TuningSet) (func() time.Duration, error) {
+	if ts.QpsLoad != nil {
+		return func() time.Duration { return time.Duration(int(float64(time.Second) / ts.QpsLoad.Qps)) }, nil
```
[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/ticker/timer.go]: done

File: clusterloader2/pkg/ticker/timer.go (hunk 1 above)
Comment (krzysied) [reply clusterloader2/pkg/ticker/timer.go]: Comments added.

File: clusterloader2/pkg/ticker/timer.go [hunk 4]
```diff
@@ -17,18 +17,67 @@ limitations under the License.
...
+		}
+		t.Reset(getDelay())
 	})
-	return t
+	return t, nil
+}
+
+// getTicksFunc returns function that specifies number of ticks that should be provided after delay.
+func getTicksFunc(ts *api.TuningSet) (func() int32, error) {
+	// for QpsLoad and RandomizedLoad, there is always 1 tick.
```
Comment (shyamjvs) [inline clusterloader2/pkg/ticker/timer.go]: Try to start comments with capital letters wherever possible.

Comment (shyamjvs) [reply clusterloader2/pkg/ticker/timer.go]: Also, maybe `For QpsLoad and RandomizedLoad, one tick is created at a time`?

File: clusterloader2/pkg/ticker/timer.go [hunk 5]
```diff
@@ -17,18 +17,67 @@ limitations under the License.
...
+	if ts.QpsLoad != nil || ts.RandomizedLoad != nil {
+		return func() int32 { return int32(1) }, nil
+	}
+	// for SteppedLoad, there is number of ticks equaled to BurstSize.
```
Comment (shyamjvs) [inline clusterloader2/pkg/ticker/timer.go]: Maybe `For SteppedLoad, 'BurstSize' many ticks are created at a time`?

File: clusterloader2/pkg/ticker/timer.go (hunk 4 above)
Comment (krzysied) [reply clusterloader2/pkg/ticker/timer.go]: Done.
I've also fixed comments. Now comments start with capital letter unless they refer to specific function name.

File: clusterloader2/pkg/ticker/timer.go (hunk 5 above)
Comment (krzysied) [reply clusterloader2/pkg/ticker/timer.go]: done

**Pull request merged**